contracts/utils/MultiRateLimited.sol

```diff
@@ -0,0 +1,203 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.4;
+
+import "../refs/CoreRef.sol";
+import "./RateLimited.sol";
+import "./IMultiRateLimited.sol";
+import "@openzeppelin/contracts/utils/math/Math.sol";
+import "@openzeppelin/contracts/utils/math/SafeCast.sol";
+
+/// @title contract for putting a rate limit on how fast an address can perform an action e.g. Minting
+/// there are two buffers, one buffer which is each individual addresses's current buffer,
+/// and then there is a global buffer which is the buffer that each individual address must respect as well
+/// @author Fei Protocol
+contract MultiRateLimited is RateLimited, IMultiRateLimited {
+    using SafeCast for *;
+
+    /// @notice the struct containing all information per rate limited address
+    struct RateLimitData {
+        uint32 lastBufferUsedTime;
+        uint144 bufferCap;
+        uint112 rateLimitPerSecond;
+        uint144 bufferStored;
```

Comment:
Can give some breathing room to some of the other vars given this is packed into 2 slots anyway

Reply:
Whatever we do here, it's cleaner to keep bufferCap and bufferStored the same type. I'm also not seeing of any scenario where we're going to replenish a buffer at a speed faster than 2^112 or have a bufferCap larger than 2^144